src/rsz/src/RepairDesign.cc

```diff
@@ -359,6 +406,243 @@ void RepairDesign::repairNet(Net* net,
+              return bufferCin(a) < bufferCin(b);
+            });
+}
+
+bool RepairDesign::performGainBuffering(Net* net,
+                                        const Pin* drvr_pin,
+                                        int max_fanout)
+{
+  struct EnqueuedPin
+  {
+    Pin* pin;
+    PathRef required_path;
+    Delay required_delay;
+    int level;
+
+    Required required(const StaState* sta) const
+    {
+      if (required_path.isNull()) {
+        return INF;
+      }
+      return required_path.required(sta) - required_delay;
+    }
+
+    std::pair<Required, int> sort_label(const StaState* sta) const
+    {
+      return std::make_pair(required(sta), -level);
+    }
+
+    float capacitance(const Network* network)
+    {
+      return network->libertyPort(pin)->capacitance();
+    }
+  };
+
+  class PinRequiredHigher
+  {
+   private:
+    const Network* network_;
+
+   public:
+    PinRequiredHigher(const Network* network) : network_(network){};
+
+    bool operator()(const EnqueuedPin& a, const EnqueuedPin& b) const
+    {
+      auto la = a.sort_label(network_), lb = b.sort_label(network_);
+      if (la > lb) {
+        return true;
+      }
+      if (la < lb) {
+        return false;
+      }
+      return sta::stringLess(network_->pathName(a.pin),
+                             network_->pathName(b.pin));
+    }
+  };
+
+  // Collect all sinks
+  std::vector<EnqueuedPin> sinks;
+
+  NetConnectedPinIterator* pin_iter = network_->connectedPinIterator(net);
+  while (pin_iter->hasNext()) {
+    const Pin* pin = pin_iter->next();
+    if (pin != drvr_pin && !network_->isTopLevelPort(pin)
+        && network_->direction(pin) == sta::PortDirection::input()
+        && network_->libertyPort(pin)) {
+      Instance* inst = network_->instance(pin);
+      if (!resizer_->dontTouch(inst)) {
+        Vertex* vertex = graph_->pinLoadVertex(pin);
+        PathRef req_path
+            = sta_->vertexWorstSlackPath(vertex, sta::MinMax::max());
+        sinks.push_back({const_cast<Pin*>(pin), req_path, 0.0, 0});
+      } else {
+        logger_->warn(RSZ,
+                      143,
+                      "dont-touch instance {} ignored when buffering net {}",
+                      network_->name(inst),
+                      network_->name(net));
+      }
+    }
+  }
+  delete pin_iter;
+
+  // Keep track of the vertices at the boundary of the tree so we know where
+  // to ask for delays to be recomputed
+  std::vector<Vertex*> tree_boundary;
+
+  float cin;
+  bool repaired_net = false;
+  if (getCin(drvr_pin, cin)) {
+    float load = 0.0;
+    for (auto& sink : sinks) {
+      load += sink.capacitance(network_);
+    }
+    std::sort(sinks.begin(), sinks.end(), PinRequiredHigher(network_));
+
+    // Iterate until we satisfy both the gain condition and max_fanout
+    // on drvr_pin
+    while (sinks.size() > max_fanout || load > cin * buffer_gain_) {
+      float max_load = bufferCin(buffer_sizes_.back()) * buffer_gain_;
```

Comment:
Can this be hoisted out of the loop / made const?

Reply:
Yes, I have pulled it out and renamed to `max_buf_load` (since it's the maximum load on a buffer, not on the driver pin at the root of the buffer tree)